### PR TITLE
Reorganize dex solver config loading

### PR DESCRIPTION
**Pull request by MartinquaXD**
Reorganized the directory structure responsible for loading configs for dex solvers to be more in line with the file structure of their logic.
This makes it so that config loading and the solver implementation are composed of 2 parts.
1. logic that is shared across all dex solvers
2. logic that only relates to each solver's dex API

Also adds a new argument which controls how we handle partially fillable orders required for [this](https://github.com/cowprotocol/services/pull/1358#issuecomment-1479201711).

### Test Plan
CI

### CODE REVIEW

File: crates/solvers/src/infra/config/dex/mod.rs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+pub mod balancer;
+mod file;
+pub mod zeroex;
+
+use crate::domain::{dex::slippage, eth};
+
+pub struct BaseConfig {
```
Comment (nlordell) [inline crates/solvers/src/infra/config/dex/mod.rs:7]: Mostly an organizational/naming nit, but to me this is not a "base configuration" but the "dex-solver configuration" (i.e. the configuration options required by the `solver::Dex` and unrelated to the actual DEX API being used).

Maybe it makes sense to have:
```
infra::config::dex
├── solver
├── balancer
└── zeroex
```

File: crates/solvers/src/infra/config/dex/balancer/file.rs [hunk 2]
```diff
@@ -1,20 +1,17 @@
...
     serde::Deserialize,
     serde_with::serde_as,
     std::path::Path,
-    tokio::fs,
 };
 
 #[serde_as]
 #[derive(Deserialize)]
-#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+#[serde(rename_all = "kebab-case")]
```
Comment (nlordell) [inline crates/solvers/src/infra/config/dex/balancer/file.rs:14]: I liked the `deny_unknown_fields` as it helps prevents misconfigurations (imagine you had an _optional_ field `foo` and you instead set `fou`, you might silently run thinking you have `foo` configured but in reality you do not).

You could instead flatten the fields here and have something like:
```rust
#[derive(Deserialize)]
#[serde(rename_all = "kebab-case", deny_unknown_fields)]
struct balancer::Config {
    // Balancer related fields.
    endpoint: reqwest::Url,
    vault: Option<H160>,
    settlement: Option<H160>,
    
    #[flatten]
    solver: infra::config::dex::solver::File,
}
```

And then `impl From<infra::config::dex::solver::File> for <infra::config::dex::solver::Config>`.

The unfortunate thing is that it leaks the `File` type, but you can scope that with `pub(super)` to limit it to the `infra::config::dex`.

File: crates/solvers/src/infra/config/dex/file.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    /// The amount of eth a partially fillable order should be filled for at
+    /// least.
+    #[serde(default = "default_smallest_partial_fill")]
+    smallest_partial_fill: eth::U256,
+}
+
+fn default_endpoint() -> reqwest::Url {
+    "https://api.0x.org/swap/v1/".parse().unwrap()
+}
+
```
Comment (nlordell) [inline crates/solvers/src/infra/config/dex/file.rs:39]:
```suggestion
```

File: crates/solvers/src/infra/config/dex/file.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+fn default_affiliate() -> H160 {
+    contracts::Contracts::for_chain(eth::ChainId::Mainnet)
+        .settlement
+        .0
+}
+
```
Comment (nlordell) [inline crates/solvers/src/infra/config/dex/file.rs:45]:
```suggestion
```

**Review by nlordell — APPROVED** [review]
LGTM. A couple of nits:
- naming nit: "base config" is a bit vague - to me its really the configuration for DEX single-order solvers, so `solver::Config` makes a bit more sense
- I would prefer to keep `deny_unknown_fields` as it helps catch misconfiguration errors.

File: crates/solvers/src/infra/config/dex/balancer/file.rs (hunk 2 above)
Comment (nlordell) [reply crates/solvers/src/infra/config/dex/balancer/file.rs:14]: Alternatively, we can add nesting and then we don't need to flatten (and would essentially just change the format).

So:
```toml
endpoint = "https://my.endpoint"

[solver]
relative-slippage = "1.2"
```

Comment (nlordell) [reply crates/solvers/src/infra/config/dex/balancer/file.rs]: Nice, I like how it turned out.

File: crates/solvers/src/infra/config/dex/mod.rs (hunk 1 above)
Comment (nlordell) [reply crates/solvers/src/infra/config/dex/mod.rs]: Makes sense. I think that the new way you organized the configs makes a lot of sense.

**Review by fedgiac — APPROVED** [review]
Looks good to me.

**Pull request merged**